File: TestDPLS/app/src/main/java/ru/bolid/testdpls/ble/DplsModels.kt
package ru.bolid.testdpls.ble

import java.util.Calendar

enum class DplsMode(
    val wire: Int,
    val title: String,
    val dangerous: Boolean,
    /** Короткое пояснение: какой порт затрагивается. */
    val portHint: String = "",
    /** Ожидаемая реакция контроллера С2000-КДЛ (ТЗ 7.4.2). */
    val controllerEffect: String = "",
) {
    NORMAL(0, "Норма", false, "Штатное прохождение линии"),
    OPEN_T(1, "Обрыв +Т", true, "Ответвление +Т", "КДЛ: потеря устройств ответвления"),
    OPEN_MAIN(2, "Обрыв магистрали", true, "Магистраль +1 ↔ +2", "КДЛ: «Нет связи» с устройствами за разрывом"),
    SHORT_1(3, "КЗ +1", true, "Порт +1", "КДЛ: «Короткое замыкание ДПЛС»"),
    SHORT_2(4, "КЗ +2", true, "Порт +2", "КДЛ: «Короткое замыкание ДПЛС»"),
    SHORT_T(5, "КЗ +Т", true, "Ответвление +Т", "КДЛ: «Короткое замыкание ДПЛС»");

    companion object {
        fun fromWire(value: Int) = entries.firstOrNull { it.wire == value }
    }
}

enum class PowerSource(val title: String) { DPLS("ДПЛС"), RESERVE("Резерв") }

enum class RssiQuality(val title: String) {
    UNKNOWN("Нет данных"),
    GOOD("Хорошая связь"),
    MEDIUM("Средняя связь"),
    WEAK("Слабая связь"),
}

fun rssiQuality(rssi: Int?): RssiQuality = when {
    rssi == null -> RssiQuality.UNKNOWN
    rssi >= -65 -> RssiQuality.GOOD
    rssi >= -80 -> RssiQuality.MEDIUM
    else -> RssiQuality.WEAK
}

/** Пароль по ТЗ: не менее 8 символов, только латинские буквы и цифры. */
fun isValidDplsPassword(value: CharSequence): Boolean =
    value.length >= 8 && value.all { it in 'A'..'Z' || it in 'a'..'z' || it in '0'..'9' }

/**
 * Отметка времени записи журнала. Для событий текущего запуска (sequence не
 * раньше последней загрузки) есть календарные дата/время — момент загрузки
 * синхронизирован с часами телефона. Для прошлых запусков момент загрузки
 * неизвестен, поэтому показываем относительное время «от запуска» без выдуманной
 * даты (ТЗ 7.5.2).
 */
data class DplsEventTime(val dateLabel: String?, val time: String, val full: String)

fun dplsEventTime(e: EventRecord, currentRunFirstSeq: Long, bootEpochSec: Long?): DplsEventTime {
    if (e.sequence >= currentRunFirstSeq && bootEpochSec != null) {
        val cal = Calendar.getInstance().apply { timeInMillis = (bootEpochSec + e.timestampSeconds) * 1000L }
        val d = cal.get(Calendar.DAY_OF_MONTH)
        val mo = cal.get(Calendar.MONTH) + 1
        val y = cal.get(Calendar.YEAR)
        val h = cal.get(Calendar.HOUR_OF_DAY)
        val mi = cal.get(Calendar.MINUTE)
        val s = cal.get(Calendar.SECOND)
        val time = "%02d:%02d:%02d".format(h, mi, s)
        return DplsEventTime(
            dateLabel = "%02d.%02d.%04d".format(d, mo, y),
            time = time,
            full = "%02d.%02d.%04d %s".format(d, mo, y, time),
        )
    }
    val t = e.timestampSeconds
    val rel = "+%02d:%02d:%02d".format(t / 3600, (t % 3600) / 60, t % 60)
    return DplsEventTime(dateLabel = null, time = rel, full = "$rel (от запуска)")
}

/** Человекочитаемое название события журнала — общее для экрана и выгрузки. */
fun dplsEventTitle(type: Int, parameter: Int): String = when (type) {
    1 -> "Запуск устройства"
    2 -> "BLE подключение"
    3 -> "BLE отключение"
    4 -> "Успешный вход"
    5 -> "Ошибка входа · попытка $parameter"
    6 -> "Вход заблокирован"
    7 -> "Режим: ${DplsMode.fromWire(parameter)?.title ?: "код $parameter"}"
    8 -> dplsAutoReturnTitle(parameter)
    9 -> "Идентификация начата"
    10 -> "Идентификация остановлена"
    11 -> "Пароль установлен"
    12 -> "Питание: ${if (parameter == 0) "от ДПЛС" else "от резерва"}"
    13 -> "Резерв: ${if (parameter == 0) "норма" else "низкий заряд"}"
    14 -> "Автоизоляция КЗ: ${if (parameter == 0) "снята" else "активна"}"
    else -> "Событие $type · $parameter"
}

private fun dplsAutoReturnTitle(reason: Int): String = when (reason) {
    0 -> "Автовозврат в «Норма» (команда оператора)"
    1 -> "Автовозврат в «Норма» (таймер)"
    2 -> "Автовозврат в «Норма» (таймаут сессии)"
    3 -> "Автовозврат в «Норма» (отключение BLE)"
    4 -> "Автовозврат в «Норма» (низкий резерв)"
    5 -> "Автовозврат в «Норма» (ошибка)"
    6 -> "Автовозврат в «Норма» (перезапуск)"
    7 -> "Автовозврат в «Норма» (автоизоляция КЗ)"
    else -> "Автовозврат в «Норма»"
}

enum class ConnectionPhase {
    IDLE, SCANNING, CONNECTING, PAIRING, NEGOTIATING_MTU, DISCOVERING,
    SUBSCRIBING, AUTHENTICATING, SYNCHRONIZING, READY, RECONNECTING, ERROR,
}

data class DiscoveredDevice(
    val address: String,
    val advertisedName: String,
    val userName: String?,
    val deviceId: Long?,
    val rssi: Int,
    /** Advertising текущей прошивки не передаёт этот признак, поэтому null = неизвестно до HELLO. */
    val initialized: Boolean? = null,
)

data class DeviceState(
    val mode: DplsMode,
    val voltageMv: Int,
    val powerSource: PowerSource,
    val reserveLow: Boolean,
    /** Hardware is isolating a real downstream short circuit (BRIZ-T function). */
    val realShort: Boolean,
    val automaticReturnSeconds: Int,
    val uptimeSeconds: Long,
    val revision: Long,
    /** Wall-clock millis when this snapshot was received from the device. */
    val receivedAtMillis: Long = 0L,
    /** STATE_REPORT validity mask. */
    val lineVoltageValid: Boolean = true,
    val reserveValid: Boolean = true,
    val powerValid: Boolean = true,
    val autoIsoValid: Boolean = true,
    val adcCalibrated: Boolean = false,
    val port1VoltageMv: Int = voltageMv,
    val port2VoltageMv: Int = 0,
    val portTVoltageMv: Int = 0,
    val reserveVoltageMv: Int = 0,
    val port1VoltageValid: Boolean = lineVoltageValid,
    val port2VoltageValid: Boolean = false,
    val portTVoltageValid: Boolean = false,
    val reserveVoltageValid: Boolean = false,
)

data class EventRecord(
    val sequence: Long,
    val timestampSeconds: Long,
    val type: Int,
    val parameter: Int,
)

data class DeviceInfo(
    val deviceId: Long,
    val protocolVersion: Int,
    val firmwareVersion: String,
    val hardwareRevision: Int,
    val adcPresent: Boolean,
    val hardwareReadback: Boolean,
    val adcCalibrated: Boolean,
    val userName: String,
    val multiVoltageReport: Boolean = false,
) {
    val shortId: String get() = "DPLS-%08X".format(deviceId)
}

enum class SettingsOp { NONE, IN_PROGRESS, DONE, FAILED }

/** UTF-8 truncation without splitting a multi-byte code point. */
internal fun utf8Truncate(value: String, maxBytes: Int): ByteArray {
    var end = 0
    var bytes = 0
    while (end < value.length) {
        val next = if (Character.isHighSurrogate(value[end]) && end + 1 < value.length) end + 2 else end + 1
        val step = value.substring(end, next).encodeToByteArray().size
        if (bytes + step > maxBytes) break
        bytes += step
        end = next
    }
    return value.substring(0, end).encodeToByteArray()
}

data class DplsUiState(
    val phase: ConnectionPhase = ConnectionPhase.IDLE,
    val statusText: String = "Готово к поиску",
    val devices: List<DiscoveredDevice> = emptyList(),
    val selectedDevice: DiscoveredDevice? = null,
    val initialized: Boolean = false,
    val credentialsReady: Boolean = false,
    val authenticated: Boolean = false,
    val state: DeviceState? = null,
    val pendingMode: DplsMode? = null,
    val commandInProgress: Boolean = false,
    val staleState: Boolean = false,
    val lastAckMillis: Long? = null,
    val eventLog: List<EventRecord> = emptyList(),
    val deviceBootEpochSeconds: Long? = null,
    val logProgress: Float? = null,
    val identifyActive: Boolean = false,
    val identifyLedLive: Boolean = false,
    val identifyRemainingSeconds: Int = 0,
    val setupName: String = "",
    val setupPassword: String = "",
    val setupRepeatPassword: String = "",
    val awaitingUserPassword: Boolean = false,
    val deviceInfo: DeviceInfo? = null,
    val settingsOp: SettingsOp = SettingsOp.NONE,
    val settingsError: String? = null,
    val localValidationError: String? = null,
    val authLockoutSeconds: Int = 0,
    val connectionRssi: Int? = null,
    val connectionRssiUpdatedAtMillis: Long? = null,
    val permissionRecoveryRequired: Boolean = false,
    val error: String? = null,
) {
    val controlsEnabled: Boolean
        get() = phase == ConnectionPhase.READY && authenticated && !commandInProgress

    val passwordInputEnabled: Boolean
        get() = authLockoutSeconds <= 0

    val setupFormReady: Boolean
        get() = credentialsReady &&
            isValidDplsPassword(setupPassword) &&
            (initialized || (setupRepeatPassword == setupPassword && setupName.isNotBlank()))
}

File: TestDPLS-iOS/TestDPLS/BLE/DplsModels.swift
import Foundation

enum DplsMode: Int, CaseIterable, Identifiable {
    case normal = 0
    case openT = 1
    case openMain = 2
    case short1 = 3
    case short2 = 4
    case shortT = 5

    var id: Int { rawValue }

    var title: String {
        switch self {
        case .normal: return "Норма"
        case .openT: return "Обрыв +Т"
        case .openMain: return "Обрыв магистрали"
        case .short1: return "КЗ +1"
        case .short2: return "КЗ +2"
        case .shortT: return "КЗ +Т"
        }
    }

    var dangerous: Bool { self != .normal }

    var portHint: String {
        switch self {
        case .normal: return "Штатное прохождение линии"
        case .openT: return "Ответвление +Т"
        case .openMain: return "Магистраль +1 ↔ +2"
        case .short1: return "Порт +1"
        case .short2: return "Порт +2"
        case .shortT: return "Ответвление +Т"
        }
    }

    var controllerEffect: String {
        switch self {
        case .normal: return ""
        case .openT: return "КДЛ: потеря устройств ответвления"
        case .openMain: return "КДЛ: «Нет связи» с устройствами за разрывом"
        case .short1, .short2, .shortT: return "КДЛ: «Короткое замыкание ДПЛС»"
        }
    }

    static func fromWire(_ value: Int) -> DplsMode? { DplsMode(rawValue: value) }
}

enum PowerSource {
    case dpls, reserve
    var title: String { self == .dpls ? "ДПЛС" : "Резерв" }
}

struct DplsEventTime {
    let dateLabel: String?
    let time: String
    let full: String
}

func dplsEventTime(_ e: EventRecord, currentRunFirstSeq: UInt32, bootEpochSec: Int64?) -> DplsEventTime {
    if e.sequence >= currentRunFirstSeq, let boot = bootEpochSec {
        let date = Date(timeIntervalSince1970: TimeInterval(boot + Int64(e.timestampSeconds)))
        let cal = Calendar.current
        let c = cal.dateComponents([.day, .month, .year, .hour, .minute, .second], from: date)
        let time = String(format: "%02d:%02d:%02d", c.hour ?? 0, c.minute ?? 0, c.second ?? 0)
        let dateLabel = String(format: "%02d.%02d.%04d", c.day ?? 0, c.month ?? 0, c.year ?? 0)
        return DplsEventTime(dateLabel: dateLabel, time: time, full: "\(dateLabel) \(time)")
    }
    let t = e.timestampSeconds
    let rel = String(format: "+%02d:%02d:%02d", t / 3600, (t % 3600) / 60, t % 60)
    return DplsEventTime(dateLabel: nil, time: rel, full: "\(rel) (от запуска)")
}

func dplsEventTitle(type: Int, parameter: Int) -> String {
    switch type {
    case 1: return "Запуск устройства"
    case 2: return "BLE подключение"
    case 3: return "BLE отключение"
    case 4: return "Успешный вход"
    case 5: return "Ошибка входа · попытка \(parameter)"
    case 6: return "Вход заблокирован"
    case 7: return "Режим: \(DplsMode.fromWire(parameter)?.title ?? "код \(parameter)")"
    case 8: return dplsAutoReturnTitle(parameter)
    case 9: return "Идентификация начата"
    case 10: return "Идентификация остановлена"
    case 11: return "Пароль установлен"
    case 12: return "Питание: \(parameter == 0 ? "от ДПЛС" : "от резерва")"
    case 13: return "Резерв: \(parameter == 0 ? "норма" : "низкий заряд")"
    case 14: return "Автоизоляция КЗ: \(parameter == 0 ? "снята" : "активна")"
    default: return "Событие \(type) · \(parameter)"
    }
}

private func dplsAutoReturnTitle(_ reason: Int) -> String {
    switch reason {
    case 0: return "Автовозврат в «Норма» (команда оператора)"
    case 1: return "Автовозврат в «Норма» (таймер)"
    case 2: return "Автовозврат в «Норма» (таймаут сессии)"
    case 3: return "Автовозврат в «Норма» (отключение BLE)"
    case 4: return "Автовозврат в «Норма» (низкий резерв)"
    case 5: return "Автовозврат в «Норма» (ошибка)"
    case 6: return "Автовозврат в «Норма» (перезапуск)"
    case 7: return "Автовозврат в «Норма» (автоизоляция КЗ)"
    default: return "Автовозврат в «Норма»"
    }
}

enum ConnectionPhase: String {
    case idle, scanning, connecting, pairing, negotiatingMtu, discovering
    case subscribing, authenticating, synchronizing, ready, reconnecting, error
}

struct DiscoveredDevice: Identifiable, Equatable {
    /// iOS CoreBluetooth peripheral UUID (MAC недоступен).
    var id: String { address }
    let address: String
    let advertisedName: String
    var userName: String?
    var deviceId: UInt32?
    var rssi: Int
}

struct DeviceState: Equatable {
    var mode: DplsMode
    var voltageMv: Int
    var powerSource: PowerSource
    var reserveLow: Bool
    var realShort: Bool
    var automaticReturnSeconds: Int
    var uptimeSeconds: UInt32
    var revision: UInt32
    var receivedAtMillis: Int64 = 0
    var lineVoltageValid: Bool = true
    var reserveValid: Bool = true
    var powerValid: Bool = true
    var autoIsoValid: Bool = true
    var adcCalibrated: Bool = false
}

struct EventRecord: Identifiable, Equatable {
    var id: UInt32 { sequence }
    let sequence: UInt32
    let timestampSeconds: UInt32
    let type: Int
    let parameter: Int
}

struct DeviceInfo: Equatable {
    let deviceId: UInt32
    let protocolVersion: Int
    let firmwareVersion: String
    let hardwareRevision: Int
    let adcPresent: Bool
    let hardwareReadback: Bool
    let adcCalibrated: Bool
    let userName: String

    var shortId: String { String(format: "DPLS-%08X", deviceId) }
}

enum SettingsOp {
    case none, inProgress, done, failed
}

func utf8Truncate(_ value: String, maxBytes: Int) -> Data {
    var result = Data()
    for scalar in value.unicodeScalars {
        let piece = String(scalar).data(using: .utf8) ?? Data()
        if result.count + piece.count > maxBytes { break }
        result.append(piece)
    }
    return result
}

struct DplsUiState: Equatable {
    var phase: ConnectionPhase = .idle
    var statusText: String = "Готово к поиску"
    var devices: [DiscoveredDevice] = []
    var selectedDevice: DiscoveredDevice?
    var initialized: Bool = false
    var credentialsReady: Bool = false
    var authenticated: Bool = false
    var state: DeviceState?
    var pendingMode: DplsMode?
    var commandInProgress: Bool = false
    var staleState: Bool = false
    var lastAckMillis: Int64?
    var eventLog: [EventRecord] = []
    var deviceBootEpochSeconds: Int64?
    var logProgress: Float?
    var identifyActive: Bool = false
    var identifyLedLive: Bool = false
    var setupName: String = ""
    var setupPassword: String = ""
    var setupRepeatPassword: String = ""
    var awaitingUserPassword: Bool = false
    var deviceInfo: DeviceInfo?
    var settingsOp: SettingsOp = .none
    var settingsError: String?
    var error: String?

    var controlsEnabled: Bool {
        phase == .ready && authenticated && !commandInProgress
    }

    var setupFormReady: Bool {
        credentialsReady &&
            setupPassword.count >= 8 &&
            (initialized || (setupRepeatPassword == setupPassword && !setupName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty))
    }
}
